File: 104. Maximum Depth of Binary Tree.py
# Problem Link: https://leetcode.com/problems/maximum-depth-of-binary-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traverse(self, node, height, current_maximum):
        if not node:
            return max(height, current_maximum)
        return max(self.traverse(node.left, height + 1, current_maximum), self.traverse(node.right, height + 1, current_maximum))
        
            
    def maxDepth(self, root: TreeNode) -> int:
        return self.traverse(root, 0, 0)

File: Leetcode/104. Maximum Depth of Binary Tree.java
// Problem Link: https://leetcode.com/problems/maximum-depth-of-binary-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int dfs(TreeNode root, int depth) {
        if (root == null) return depth;
        return Math.max(dfs(root.left, depth + 1), dfs(root.right, depth + 1));
    }

    public int maxDepth(TreeNode root) {
        return dfs(root, 0);
    }
}
